File: complexity_check_simple.py
#!/usr/bin/env pypy3
"""Simple complexity check for Python implementation"""

import time
import math
from disk_tree import Circle, DiskStabbingDS
import random

def count_nodes(node):
    """Count nodes in segment tree"""
    if node is None:
        return 0
    return 1 + count_nodes(node.left) + count_nodes(node.right)

def main():
    print()
    print("Simple Complexity Check (Python)")
    print("=" * 60)
    print("Checking if query time ≈ O(log n) and space ≈ O(n log n)\n")

    sizes = [1000, 2000, 4000, 8000]

    prev_n = 0
    prev_time = 0
    prev_space = 0

    for i, n in enumerate(sizes):
        random.seed(42)

        circles = [
            Circle(
                (random.random() - 0.5) * 1000.0,
                (random.random() - 0.5) * 1000.0,
                random.random() * 50.0 + 1.0
            )
            for _ in range(n)
        ]

        ds = DiskStabbingDS(circles)
        ds.build()

        # Measure query time
        start = time.perf_counter()
        for _ in range(1000):
            xq = (random.random() - 0.5) * 1100.0
            yq = (random.random() - 0.5) * 1100.0
            ds.query(xq, yq)
        elapsed_us = (time.perf_counter() - start) * 1_000_000

        # Count nodes
        nodes = sum(count_nodes(slab.root) for slab in ds._slabs)

        log_n = math.log(n)
        n_logn = n * log_n

        print(f"n={n:5}  time={elapsed_us:6.1f}µs  nodes={nodes:7}  "
              f"log(n)={log_n:.2f}  n*log(n)={n_logn:.0f}")

        if i > 0:
            n_ratio = n / prev_n
            time_ratio = elapsed_us / prev_time
            space_ratio = nodes / prev_space

            expected_time = math.log(n) / math.log(prev_n)
            expected_space = n_logn / (prev_n * math.log(prev_n))

            print(f"  → n grew {n_ratio:.1f}x, time grew {time_ratio:.2f}x "
                  f"(expect {expected_time:.2f}x for log), "
                  f"space grew {space_ratio:.2f}x (expect {expected_space:.2f}x for n log n)")

            time_ok = abs(time_ratio - expected_time) < expected_time
            space_ok = abs(space_ratio - expected_space) < expected_space * 0.5

            if time_ok:
                print("     ✓ Time scaling looks like O(log n)")
            else:
                print("     ⚠ Time scaling off from O(log n)")

            if space_ok:
                print("     ✓ Space scaling looks like O(n log n)")
            else:
                print("     ⚠ Space scaling off from O(n log n)")

        print()

        prev_n = n
        prev_time = elapsed_us
        prev_space = nodes

    print("=" * 60)
    print("If most checks show ✓, complexities are good!")
    print()

if __name__ == "__main__":
    main()

File: complexity_check_simple.zig
const std = @import("std");
const DiskStabbingDS = @import("disk_stabbing.zig").DiskStabbingDS;
const Circle = @import("disk_stabbing.zig").Circle;
const SegNode = @import("disk_stabbing.zig").SegNode;

fn countNodes(node: ?*SegNode) u64 {
    if (node) |n| {
        return 1 + countNodes(n.left) + countNodes(n.right);
    }
    return 0;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("\n", .{});
    std.debug.print("Simple Complexity Check\n", .{});
    std.debug.print("=" ** 60 ++ "\n", .{});
    std.debug.print("Checking if query time ≈ O(log n) and space ≈ O(n log n)\n\n", .{});

    const sizes = [_]usize{ 1000, 2000, 4000, 8000 };

    var prev_n: f64 = 0;
    var prev_time: f64 = 0;
    var prev_space: f64 = 0;

    for (sizes, 0..) |n, i| {
        var prng = std.Random.DefaultPrng.init(42);
        var rng = prng.random();

        const circles = try allocator.alloc(Circle, n);
        defer allocator.free(circles);

        for (circles) |*c| {
            c.cx = (rng.float(f64) - 0.5) * 1000.0;
            c.cy = (rng.float(f64) - 0.5) * 1000.0;
            c.r = rng.float(f64) * 50.0 + 1.0;
        }

        var ds = try DiskStabbingDS.init(allocator, circles);
        defer ds.deinit();
        try ds.build();

        // Measure query time
        const start = std.time.nanoTimestamp();
        for (0..1000) |_| {
            const xq = (rng.float(f64) - 0.5) * 1100.0;
            const yq = (rng.float(f64) - 0.5) * 1100.0;
            const result = try ds.query(xq, yq, allocator);
            allocator.free(result);
        }
        const elapsed_us = @as(f64, @floatFromInt(std.time.nanoTimestamp() - start)) / 1000.0 / 1000.0;

        // Count nodes
        var nodes: u64 = 0;
        for (ds.slabs) |slab| {
            if (slab.root) |root| {
                nodes += countNodes(root);
            }
        }

        const n_f = @as(f64, @floatFromInt(n));
        const log_n = @log(n_f);
        const n_logn = n_f * log_n;

        std.debug.print("n={d:5}  time={d:6.1}µs  nodes={d:7}  log(n)={d:.2}  n*log(n)={d:.0}\n", .{
            n,
            elapsed_us,
            nodes,
            log_n,
            n_logn,
        });

        if (i > 0) {
            const n_ratio = n_f / prev_n;
            const time_ratio = elapsed_us / prev_time;
            const space_ratio = @as(f64, @floatFromInt(nodes)) / prev_space;

            const expected_time_ratio = @log(n_f) / @log(prev_n); // log ratio
            const expected_space_ratio = n_logn / (prev_n * @log(prev_n)); // n log n ratio

            std.debug.print("  → n grew {d:.1}x, time grew {d:.2}x (expect {d:.2}x for log), space grew {d:.2}x (expect {d:.2}x for n log n)\n", .{
                n_ratio,
                time_ratio,
                expected_time_ratio,
                space_ratio,
                expected_space_ratio,
            });

            // Simple check: ratios should be within 2x of expected
            const time_ok = @abs(time_ratio - expected_time_ratio) < expected_time_ratio;
            const space_ok = @abs(space_ratio - expected_space_ratio) < expected_space_ratio * 0.5;

            if (time_ok) {
                std.debug.print("     ✓ Time scaling looks like O(log n)\n", .{});
            } else {
                std.debug.print("     ⚠ Time scaling off from O(log n)\n", .{});
            }

            if (space_ok) {
                std.debug.print("     ✓ Space scaling looks like O(n log n)\n", .{});
            } else {
                std.debug.print("     ⚠ Space scaling off from O(n log n)\n", .{});
            }
        }
        std.debug.print("\n", .{});

        prev_n = n_f;
        prev_time = elapsed_us;
        prev_space = @floatFromInt(nodes);
    }

    std.debug.print("=" ** 60 ++ "\n", .{});
    std.debug.print("If most checks show ✓, complexities are good!\n", .{});
    std.debug.print("\n", .{});
}
